subsys/bluetooth/audio/has.c

```diff
@@ -0,0 +1,1330 @@
+			atomic_clear_bit(client->flags, CLIENT_FLAG_CP_BUSY);
+		}
+	}
+}
+
+static int preset_name_set(uint8_t id, const char *name, uint8_t len)
+{
+#if defined(CONFIG_BT_HAS_PRESET_NAME_DYNAMIC)
+	struct has_preset *preset;
+
+	if (len < BT_HAS_PRESET_NAME_MIN || len > BT_HAS_PRESET_NAME_MAX) {
+		return BT_GATT_ERR(BT_HAS_ERR_INVALID_PARAM_LEN);
+	}
+
+	preset = preset_get(id);
+	if (!preset) {
+		return BT_GATT_ERR(BT_ATT_ERR_OUT_OF_RANGE);
+	}
+
+	if (!(preset->properties & BT_HAS_PROP_WRITABLE)) {
+		return BT_GATT_ERR(BT_HAS_ERR_WRITE_NAME_NOT_ALLOWED);
+	}
+
+	strncpy(preset->name, name, len);
+	preset->name[len] = '\0';
+
+	/* Do not send preset changed notification if the preset is hidden */
+	if (!preset->hidden) {
+		atomic_clear(preset->is_client_name_aware);
+		preset_changed(preset, BT_HAS_CHANGE_ID_GENERIC_UPDATE);
+	}
+
+	if (preset_ops->name_changed) {
+		preset_ops->name_changed(&has_local, id, preset->name);
+	}
+
+	return 0;
+}
+
+static int handle_write_preset_name(struct has_client *client, struct net_buf_simple *buf)
+{
+	const struct bt_has_cp_write_preset_name_req *req;
+
+	if (buf->len < sizeof(*req)) {
+		return BT_GATT_ERR(BT_HAS_ERR_INVALID_PARAM_LEN);
+	}
+
+	req = net_buf_simple_pull_mem(buf, sizeof(*req));
+
+	return preset_name_set(req->idx, req->name, buf->len);
+}
+#else
+	return -EOPNOTSUPP;
+}
+#endif /* CONFIG_BT_HAS_PRESET_NAME_DYNAMIC */
+
+static int call_ops_set_active_preset(uint8_t id, bool sync)
+{
+	if (preset_ops == NULL) {
+		return BT_GATT_ERR(BT_HAS_ERR_OPERATION_NOT_POSSIBLE);
+	}
+
+	if (preset_ops->active_set(&has_local, id, sync) != 0) {
+		return BT_GATT_ERR(BT_HAS_ERR_OPERATION_NOT_POSSIBLE);
+	}
+
+	return 0;
+}
+
+static int handle_set_active_preset(struct has_client *client, struct net_buf_simple *buf,
+				    bool sync)
+{
+	const struct bt_has_cp_set_active_preset_req *req;
+	struct has_preset *preset;
+
+	if (buf->len < sizeof(*req)) {
+		return BT_GATT_ERR(BT_HAS_ERR_INVALID_PARAM_LEN);
+	}
+
+	req = net_buf_simple_pull_mem(buf, sizeof(*req));
+
+	preset = preset_get(req->idx);
+	if (!preset) {
+		return BT_GATT_ERR(BT_ATT_ERR_OUT_OF_RANGE);
+	}
+
+	if (!(preset->properties & BT_HAS_PROP_AVAILABLE)) {
+		return BT_GATT_ERR(BT_HAS_ERR_OPERATION_NOT_POSSIBLE);
+	}
+
+	return call_ops_set_active_preset(preset->id, sync);
+}
+
+static bool find_available(struct has_preset *preset, void *user_data)
+{
+	struct has_preset **found = user_data;
+
+	if ((preset->properties & BT_HAS_PROP_AVAILABLE) && !preset->hidden) {
+		*found = preset;
+		return true;
+	}
+
+	return false;
+}
+
+static int handle_set_next_preset(struct has_client *client, struct net_buf_simple *buf, bool sync)
+{
+	struct has_preset *preset = NULL;
+
+	preset_foreach(has_local.active_id + 1, last_preset_id, 1, find_available, &preset);
+	if (preset) {
+		return call_ops_set_active_preset(preset->id, sync);
+	}
+
+	preset_foreach(0x01, has_local.active_id - 1, 1, find_available, &preset);
```

Comment:
Off-topic: This I have seen many times (elswhere).  Why are numbers some times given in hex, and sometimes not?  The index is given in hex - "0x01", while `num_matches` is given as "1", in decimal.

Reply:
fixed to use decimal